File: backend_js/my_agent/DatabaseManager.ts
import axios from "axios";

export class DatabaseManager {
  private endpoint_url: string;

  constructor() {
    this.endpoint_url =
      process.env.DB_ENDPOINT_URL || "http://host.docker.internal:3001";
    this.get_schema = this.get_schema.bind(this);
    this.execute_query = this.execute_query.bind(this);
  }

  async get_schema(uuid: string): Promise<string> {
    try {
      console.log("Getting schema for uuid: ", this.endpoint_url);
      const response = await axios.get(
        `${this.endpoint_url}/get-schema/${uuid}`
      );
      return response.data.schema;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(`Error fetching schema: ${error.message}`);
      } else {
        throw new Error(`Error fetching schema: ${String(error)}`);
      }
    }
  }

  async execute_query(uuid: string, query: string): Promise<any[]> {
    try {
      const response = await axios.post(`${this.endpoint_url}/execute-query`, {
        uuid,
        query,
      });
      return response.data.results;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(`Error executing query: ${error.message}`);
      } else {
        throw new Error(`Error executing query: ${String(error)}`);
      }
    }
  }
}

File: backend_py/my_agent/DatabaseManager.py
import requests
import os
from typing import List, Any


class DatabaseManager:
    def __init__(self):
        self.endpoint_url = os.getenv("DB_ENDPOINT_URL")

    def get_schema(self, uuid: str) -> str:
        """Retrieve the database schema."""
        try:
            print(f"{self.endpoint_url}/get-schema/{uuid}")
            response = requests.get(
                f"{self.endpoint_url}/get-schema/{uuid}"
            )
            response.raise_for_status()
            return response.json()['schema']
        except requests.RequestException as e:
            raise Exception(f"Error fetching schema: {str(e)}")

    def execute_query(self, uuid: str, query: str) -> List[Any]:
        """Execute SQL query on the remote database and return results."""
        try:
            response = requests.post(
                f"{self.endpoint_url}/execute-query",
                json={"uuid": uuid, "query": query}
            )
            response.raise_for_status()
            return response.json()['results']
        except requests.RequestException as e:
            raise Exception(f"Error executing query: {str(e)}")